### ISSUE TITLE
pypi tarball missing distro/*

### ISSUE THREAD

**Issue by markrechler** [graphite-project/carbon#355]
It seems the carbon-0.9.13.tar.gz file uploaded to pypi on 2014-12-31 (MD5: 98c182967f52dd0d93a9c4a7a4066b5b) does not have the distro folder and relevant init scripts causing pip installs on redhat to bomb:

`error: can't copy 'distro/redhat/init.d/carbon-cache': doesn't exist or not a regular file`

Perhaps it was not built on a centos/redhat box.

**Comment by SEJeff** [comment]
I should be able to take a shot at this issue this week

**Comment by obfuscurity** [comment]
Thanks @SEJeff. :smile_cat:

**Comment by TinajaLabs** [comment]
I just hit this error...  Is there a fix or a reasonable work-around?

Thanks,
Chris.


results when running ```pip install carbon``` on rhel 6.5, python 2.6.6.  All seems good then as it is copying files:

```
    copying conf/carbon.conf.example -> /opt/graphite/conf

    error: can't copy 'distro/redhat/init.d/carbon-cache': doesn't exist or not a regular file

    ----------------------------------------
    Command "/usr/bin/python -c "import setuptools, tokenize;__file__='/tmp/pip-build-i4fUqf/carbon/setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" install --record /tmp/pip-9ZK2i5-record/install-record.txt --single-version-externally-managed --compile" failed with error code 1 in /tmp/pip-build-i4fUqf/carbon
```

**Comment by markrechler** [comment]
You could rebuild it on your rhel box and then install or host it locally:
```
wget 'https://pypi.python.org/packages/source/c/carbon/carbon-0.9.13.tar.gz#md5=98c182967f52dd0d93a9c4a7a4066b5b' && tar xpf carbon-0.9.13.tar.gz && cd carbon-0.9.13 && python setup.py sdist
```
The new package with the init.d scripts will be in `dist`.

**Comment by TinajaLabs** [comment]
DOH! Thanks, I got close to the same problem...  Let's see if my steps were correct...

I ran these commands as you suggested:
```
cd /opt
wget 'https://pypi.python.org/packages/source/c/carbon/carbon-0.9.13.tar.gz#md5=98c182967f52dd0d93a9c4a7a4066b5b'
tar xpf carbon-0.9.13.tar.gz
cd carbon-0.9.13
python setup.py sdist
```
It responded like this:
```
/usr/lib64/python2.6/distutils/dist.py:266: UserWarning: Unknown distribution option: 'install_requires'
  warnings.warn(msg)
running sdist
warning: sdist: standard file not found: should have one of README, README.txt
package init file 'lib/twisted/plugins/__init__.py' not found (or not a regular file)
reading manifest template 'MANIFEST.in'
warning: no files found matching '*' under directory 'distro/'
warning: no previously-included files found matching 'conf/*.conf'
writing manifest file 'MANIFEST'
creating carbon-0.9.13
creating carbon-0.9.13/bin
creating carbon-0.9.13/conf
creating carbon-0.9.13/lib
creating carbon-0.9.13/lib/carbon
creating carbon-0.9.13/lib/carbon/aggregator
creating carbon-0.9.13/lib/twisted
creating carbon-0.9.13/lib/twisted/plugins
making hard links in carbon-0.9.13...
hard linking LICENSE -> carbon-0.9.13
hard linking MANIFEST.in -> carbon-0.9.13
hard linking setup.cfg -> carbon-0.9.13
hard linking setup.py -> carbon-0.9.13
hard linking bin/carbon-aggregator.py -> carbon-0.9.13/bin
hard linking bin/carbon-cache.py -> carbon-0.9.13/bin
hard linking bin/carbon-client.py -> carbon-0.9.13/bin
hard linking bin/carbon-relay.py -> carbon-0.9.13/bin
hard linking bin/validate-storage-schemas.py -> carbon-0.9.13/bin
hard linking conf/aggregation-rules.conf.example -> carbon-0.9.13/conf
hard linking conf/blacklist.conf.example -> carbon-0.9.13/conf
hard linking conf/carbon.amqp.conf.example -> carbon-0.9.13/conf
hard linking conf/carbon.conf.example -> carbon-0.9.13/conf
hard linking conf/relay-rules.conf.example -> carbon-0.9.13/conf
hard linking conf/rewrite-rules.conf.example -> carbon-0.9.13/conf
hard linking conf/storage-aggregation.conf.example -> carbon-0.9.13/conf
hard linking conf/storage-schemas.conf.example -> carbon-0.9.13/conf
hard linking conf/whitelist.conf.example -> carbon-0.9.13/conf
hard linking lib/carbon/__init__.py -> carbon-0.9.13/lib/carbon
hard linking lib/carbon/amqp0-8.xml -> carbon-0.9.13/lib/carbon
hard linking lib/carbon/amqp_listener.py -> carbon-0.9.13/lib/carbon
hard linking lib/carbon/amqp_publisher.py -> carbon-0.9.13/lib/carbon
hard linking lib/carbon/cache.py -> carbon-0.9.13/lib/carbon
hard linking lib/carbon/client.py -> carbon-0.9.13/lib/carbon
hard linking lib/carbon/conf.py -> carbon-0.9.13/lib/carbon
hard linking lib/carbon/events.py -> carbon-0.9.13/lib/carbon
hard linking lib/carbon/exceptions.py -> carbon-0.9.13/lib/carbon
hard linking lib/carbon/hashing.py -> carbon-0.9.13/lib/carbon
hard linking lib/carbon/instrumentation.py -> carbon-0.9.13/lib/carbon
hard linking lib/carbon/log.py -> carbon-0.9.13/lib/carbon
hard linking lib/carbon/management.py -> carbon-0.9.13/lib/carbon
hard linking lib/carbon/manhole.py -> carbon-0.9.13/lib/carbon
hard linking lib/carbon/protocols.py -> carbon-0.9.13/lib/carbon
hard linking lib/carbon/regexlist.py -> carbon-0.9.13/lib/carbon
hard linking lib/carbon/relayrules.py -> carbon-0.9.13/lib/carbon
hard linking lib/carbon/rewrite.py -> carbon-0.9.13/lib/carbon
hard linking lib/carbon/routers.py -> carbon-0.9.13/lib/carbon
hard linking lib/carbon/service.py -> carbon-0.9.13/lib/carbon
hard linking lib/carbon/state.py -> carbon-0.9.13/lib/carbon
hard linking lib/carbon/storage.py -> carbon-0.9.13/lib/carbon
hard linking lib/carbon/util.py -> carbon-0.9.13/lib/carbon
hard linking lib/carbon/writer.py -> carbon-0.9.13/lib/carbon
hard linking lib/carbon/aggregator/__init__.py -> carbon-0.9.13/lib/carbon/aggregator
hard linking lib/carbon/aggregator/buffers.py -> carbon-0.9.13/lib/carbon/aggregator
hard linking lib/carbon/aggregator/receiver.py -> carbon-0.9.13/lib/carbon/aggregator
hard linking lib/carbon/aggregator/rules.py -> carbon-0.9.13/lib/carbon/aggregator
hard linking lib/twisted/plugins/carbon_aggregator_plugin.py -> carbon-0.9.13/lib/twisted/plugins
hard linking lib/twisted/plugins/carbon_cache_plugin.py -> carbon-0.9.13/lib/twisted/plugins
hard linking lib/twisted/plugins/carbon_relay_plugin.py -> carbon-0.9.13/lib/twisted/plugins
creating dist
tar -cf dist/carbon-0.9.13.tar carbon-0.9.13
gzip -f9 dist/carbon-0.9.13.tar
tar -cf dist/carbon-0.9.13.tar carbon-0.9.13
gzip -f9 dist/carbon-0.9.13.tar
removing 'carbon-0.9.13' (and everything under it)
```
There are warnings about ```/distro``` and ```/conf```
I found the dir named ```dist``` with a new tar file:  ```carbon-0.9.13.tar.gz```

I untared it as before:
```
cd dist
tar xpf carbon-0.9.13.tar.gz
cd carbon-0.9.13
python setup.py install
```

Response:
```
/usr/lib64/python2.6/distutils/dist.py:266: UserWarning: Unknown distribution option: 'install_requires'
  warnings.warn(msg)
running install
running build
running build_py
creating build
creating build/lib
creating build/lib/carbon
copying lib/carbon/amqp_listener.py -> build/lib/carbon
copying lib/carbon/amqp_publisher.py -> build/lib/carbon
copying lib/carbon/instrumentation.py -> build/lib/carbon
copying lib/carbon/__init__.py -> build/lib/carbon
copying lib/carbon/util.py -> build/lib/carbon
copying lib/carbon/cache.py -> build/lib/carbon
copying lib/carbon/exceptions.py -> build/lib/carbon
copying lib/carbon/protocols.py -> build/lib/carbon
copying lib/carbon/rewrite.py -> build/lib/carbon
copying lib/carbon/events.py -> build/lib/carbon
copying lib/carbon/storage.py -> build/lib/carbon
copying lib/carbon/manhole.py -> build/lib/carbon
copying lib/carbon/state.py -> build/lib/carbon
copying lib/carbon/regexlist.py -> build/lib/carbon
copying lib/carbon/conf.py -> build/lib/carbon
copying lib/carbon/routers.py -> build/lib/carbon
copying lib/carbon/management.py -> build/lib/carbon
copying lib/carbon/client.py -> build/lib/carbon
copying lib/carbon/writer.py -> build/lib/carbon
copying lib/carbon/hashing.py -> build/lib/carbon
copying lib/carbon/service.py -> build/lib/carbon
copying lib/carbon/log.py -> build/lib/carbon
copying lib/carbon/relayrules.py -> build/lib/carbon
creating build/lib/carbon/aggregator
copying lib/carbon/aggregator/__init__.py -> build/lib/carbon/aggregator
copying lib/carbon/aggregator/buffers.py -> build/lib/carbon/aggregator
copying lib/carbon/aggregator/rules.py -> build/lib/carbon/aggregator
copying lib/carbon/aggregator/receiver.py -> build/lib/carbon/aggregator
package init file 'lib/twisted/plugins/__init__.py' not found (or not a regular file)
creating build/lib/twisted
creating build/lib/twisted/plugins
copying lib/twisted/plugins/carbon_relay_plugin.py -> build/lib/twisted/plugins
copying lib/twisted/plugins/carbon_aggregator_plugin.py -> build/lib/twisted/plugins
copying lib/twisted/plugins/carbon_cache_plugin.py -> build/lib/twisted/plugins
copying lib/carbon/amqp0-8.xml -> build/lib/carbon
package init file 'lib/twisted/plugins/__init__.py' not found (or not a regular file)
running build_scripts
creating build/scripts-2.6
copying and adjusting bin/validate-storage-schemas.py -> build/scripts-2.6
copying and adjusting bin/carbon-aggregator.py -> build/scripts-2.6
copying and adjusting bin/carbon-relay.py -> build/scripts-2.6
copying and adjusting bin/carbon-cache.py -> build/scripts-2.6
copying and adjusting bin/carbon-client.py -> build/scripts-2.6
changing mode of build/scripts-2.6/validate-storage-schemas.py from 644 to 755
changing mode of build/scripts-2.6/carbon-aggregator.py from 644 to 755
changing mode of build/scripts-2.6/carbon-relay.py from 644 to 755
changing mode of build/scripts-2.6/carbon-cache.py from 644 to 755
changing mode of build/scripts-2.6/carbon-client.py from 644 to 755
running install_lib
running install_scripts
copying build/scripts-2.6/validate-storage-schemas.py -> /opt/graphite/bin
copying build/scripts-2.6/carbon-aggregator.py -> /opt/graphite/bin
copying build/scripts-2.6/carbon-relay.py -> /opt/graphite/bin
copying build/scripts-2.6/carbon-cache.py -> /opt/graphite/bin
copying build/scripts-2.6/carbon-client.py -> /opt/graphite/bin
changing mode of /opt/graphite/bin/validate-storage-schemas.py to 755
changing mode of /opt/graphite/bin/carbon-aggregator.py to 755
changing mode of /opt/graphite/bin/carbon-relay.py to 755
changing mode of /opt/graphite/bin/carbon-cache.py to 755
changing mode of /opt/graphite/bin/carbon-client.py to 755
running install_data
error: can't copy 'distro/redhat/init.d/carbon-cache': doesn't exist or not a regular file
```

DOH!
seems the distro/redhat/just isn't in the picture...


I had success with graphite-web and whisper so the only stumbling block seems to be carbon.  I suppose I could copy the init.d files from the git repo but I'm wondering what else is missing...  

or, am I just confused ;)

**Comment by markrechler** [comment]
Ah, you are right, the source distro files don't seem to be in the tar either.

You could do a clone and checkout the tag:
```
git clone git@github.com:graphite-project/carbon.git
cd carbon
git checkout 0.9.13-pre1
python setup.py sdist
```

**Comment by dlabreu** [comment]
Hi There guys, do you guys know a away around on that? All my servers are RHEL7 and I need to run this on RHEL7.

**Comment by ajsmith** [comment]
@dlabreu I got around this on RHEL 6 by just using the packages from EPEL. I imagine you might have the same luck with RHEL 7.

**Comment by patelgaurang83** [comment]
I am seeing same issue on RHEL. Any update on this issue? workaround?

**Comment by obfuscurity** [comment]
@SEJeff @bmhatfield Any suggestions on how to fix this? Let's get this put to rest so we can cut a release.

**Comment by deniszh** [comment]
Ahem... Maybe I'm wrong but what's the problem here? carbon-0.9.13.tar.gz from https://pypi.python.org/pypi/carbon/0.9.13 really doesn't contain distro directory. So, we need to include that in upcoming 0.9.14 release - and that's all. Right?

**Comment by obfuscurity** [comment]
@deniszh Point is, the directory and files have existed there for 3 years, so why isn't it being included in the tarball?

**Comment by deniszh** [comment]
@obfuscurity - existing where? carbon-0.9.13.tar.gz was uploaded 2014-12-31
Maybe it was some glitch in pypi during upload?

**Comment by obfuscurity** [comment]
@deniszh The stuff in https://github.com/graphite-project/carbon/tree/master/distro/redhat/init.d has been there since 2012. I don't know why it would be missing from the pypi tarball.

**Comment by bmhatfield** [comment]
Okay, I took a look and I think I found the problem. Somewhat amusingly, I believe the issue filer also identified the same problem, but they were a bit unclear about it.

In setup.py, we have this stanza:

https://github.com/graphite-project/carbon/blob/master/setup.py#L58-L64

```
# If we are building on RedHat/CentOS, let's use the redhat init scripts.
# RHEL & Scientific Linux both output 'redhat'.
if platform.dist()[0] in ('redhat', 'centos', 'fedora'):
    init_scripts = [ ('/etc/init.d', ['distro/redhat/init.d/carbon-cache',
                                      'distro/redhat/init.d/carbon-relay',
                                      'distro/redhat/init.d/carbon-aggregator']) ]
    install_files += init_scripts
```

I interpret this as meaning that when we run `python setup.py bdist`, it results in this statement being executed. If the build command is not run on a redhat box, the init scripts will not be included in the tarball. Then, when someone who *is* on a redhat box tries to install this using setup.py (or a wrapper that calls it), the expected manifest will be different than the contents of the tarball.

I'm not sure who actually uploaded the tarball for 0.9.13, but I think there's a very reasonable chance it was built on a non-redhat linux box.

There are a couple ways we could address this, as I don't think that uploading separate tarballs for redhat/ubuntu is really the right way to go.

- We could remove the init scripts entirely from setup.py, and require folks to manage this on their own.
- We could always include a distro/example folder and require folks to link in the init scripts on their own
- We could investigate some other setup.py solution such that all the init scripts end up in the tarball but only the correct one is installed at runtime. I'm not sure how we would do this, but I suspect that it would have some dependence on differentiating between `sdist` mode and `install` mode.

**Comment by obfuscurity** [comment]
Yes, I'm almost certain I built it on an Ubuntu box.

**Comment by deniszh** [comment]
IMO always including of distro/example folder is easiest solution. :+1: :smiling_imp:

**Comment by obfuscurity** [comment]
Yes, I think that makes the most sense. Anyone care to submit a PR so we can close this out and :ship:?

**Comment by deniszh** [comment]
Is https://github.com/graphite-project/carbon/pull/454 looks good? Will port to master then.

**Comment by obfuscurity** [comment]
LGTM, but I'd like to get @bmhatfield's opinion.

**Comment by obfuscurity** [comment]
Fixed by #454.
